Fleet Desktop MVP

lucasmrod opened the pull request:
#4429 & #3914

![Screen Shot 2022-03-08 at 14 05 04](https://user-images.githubusercontent.com/2073526/157423820-2f8fb126-6da5-4709-a042-8562d0029202.png)

- [x] Changes file added (for user-visible changes)
- ~[ ] Documented any API changes (docs/Using-Fleet/REST-API.md)~
- ~[ ] Documented any permissions changes~
- [x] Ensured that input data is properly validated, SQL injection is prevented (using placeholders for values in statements)
- [x] Added/updated tests
- [x] Manual QA for all new/changed functionality

chiiph commented on cmd/fleetctl/desktop.go:
```diff
@@ -0,0 +1,180 @@
+	infoFile := filepath.Join(contentsDir, "Info.plist")
+	ioutil.WriteFile(infoFile, []byte(infoPList), 0o644)
+
+	cmd := exec.Command("go", "build", "-o", filepath.Join(macOSDir, "fleet-desktop"), "./"+filepath.Join("orbit", "cmd", "desktop"))
```
This is certainly ok at this stage, but the end result shouldn't need the fleetctl user to have go installed in their path. This is probably stupidly obvious, but I felt compelled to mention it.

lucasmrod replied:
This command (`fleetctl desktop`) is to be used only by TUF maintainers. What other alternative can you think of?

chiiph replied:
Github Action with official releases? Wouldn't this be used as part of `fleetctl package` by users? In that case, it would work the same as with orbit, which I don't think we build there.

lucasmrod replied:
> Github Action with official releases?

I see. We could just use `fleetctl desktop` command in Github CI to release.
Or, I could also move it to `tools` instead. (We will eventually need facilities for people managing their TUF servers.)

> Wouldn't this be used as part of fleetctl package by users?

- `fleetctl desktop` (or a tool under `tools`) would help build the target for desktop.
- `fleetctl updates add` is used to push such target to TUF.
- `fleetctl package` fetches targets from TUF.

chiiph commented on orbit/cmd/desktop/desktop.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+//go:embed favicon.ico
+var icoBytes []byte
```
We will probably want a handful of different icons, not sure if it's worth while to think about this at this point, but eventually we'll use it to show different states such as policy not passing, or update needed, etc.

chiiph commented on orbit/cmd/desktop/desktop.go:
```diff
@@ -0,0 +1,38 @@
+var icoBytes []byte
+
+func main() {
+	// Our TUF provided targets must support launching with "--help".
```
Oh because that's part of the checks after downloading a binary?

lucasmrod replied:
Correct. Am thinking on having `"--check"` instead on all our executables that just exits successfully. (Principle of least surprise.)

chiiph commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -408,13 +446,32 @@ func main() {
 
+		if runtime.GOOS == "darwin" && c.Bool("fleet-desktop") {
+			done := make(chan struct{})
+			// TODO(lucas): Kill .app in the interrupt.
```
Do we have to? Orbit should relaunch if interrupted, and the desktop app doesn't depend on orbit except to generate the auth token file.

If the file is not present, we just disable the menu item, or show a modal with the error, and they can be completely independent.

lucasmrod replied:
Haven't thought this part through yet.

> Do we have to? Orbit should relaunch if interrupted, and the desktop app doesn't depend on orbit except to generate the auth token file.

We do if we want to allow auto-updating the "Fleet Desktop" application.

chiiph replied:
Ah, right. That way we don't have to care about that logic in desktop too... makes sense!

lucasmrod commented:
@noahtalerman FYI: See the screen capture in the description. Fleet Desktop MVP has one entry in the tray icon, "My Device".

mna commented on tools/desktop/desktop.go:
```diff
@@ -0,0 +1,230 @@
+	)
+
+	if runtime.GOOS != "darwin" {
+		return errors.New("The \"Fleet Desktop\" macOS app can only be created from macOS")
```
Very minor, but:
```suggestion
		return errors.New(`The "Fleet Desktop" macOS app can only be created from macOS`)
```

mna commented on tools/desktop/desktop.go:
```diff
@@ -0,0 +1,230 @@
+
+	contentsDir := filepath.Join(appDir, "Contents")
+	macOSDir := filepath.Join(contentsDir, "MacOS")
+	for _, dir := range []string{macOSDir} {
```
Is that because you expect to add more soon? Or should `contentsDir` be part of the slice?

mna commented on tools/desktop/desktop.go:
```diff
@@ -0,0 +1,230 @@
+	macOSDir := filepath.Join(contentsDir, "MacOS")
+	for _, dir := range []string{macOSDir} {
+		if err := secure.MkdirAll(dir, constant.DefaultDirMode); err != nil {
+			return fmt.Errorf("create directories: %w", err)
```
```suggestion
			return fmt.Errorf("create directory %q: %w", dir, err)
```

mna commented on tools/desktop/desktop.go:
```diff
@@ -0,0 +1,230 @@
+
+	infoFile := filepath.Join(contentsDir, "Info.plist")
+	infoPListContents := fmt.Sprintf(infoPList, bundleIdentifier, version, version)
+	ioutil.WriteFile(infoFile, []byte(infoPListContents), 0o644)
```
Should check the error here.

mna reviewed (commented):
Looks good, can't comment much on the orbit packaging/installing approach _per se_ as I'm not super familiar with it, but I left some comments on the code, nothing major.

zwass commented on tools/desktop/desktop.go:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	/* #nosec G204 -- arguments are actually well defined */
+	buildExec := exec.Command("go", "build", "-o", filepath.Join(macOSDir, "fleet-desktop"), "./"+filepath.Join("orbit", "cmd", "desktop"))
```
Perhaps this?
```suggestion
	buildExec := exec.Command("go", "build", "-o", filepath.Join(macOSDir, "fleet-desktop"), filepath.Join(".", "orbit", "cmd", "desktop"))
```

lucasmrod replied:
Didn't help, looks like the only thing that would work is replacing the `filepath.Join` calls with the actual strings.

zwass reviewed (commented):
The `init_tuf` and related script made it so easy to test this! Looking great overall, I don't have any concerns with the approach.

I thought the icon seemed a bit small in the menu bar so I added a slightly larger icon.

lucasmrod commented:
One thing to note: the while logo may not look great on a light background (because it's not a template icon file AFAICS):
![Screen Shot 2022-03-21 at 10 41 47](https://user-images.githubusercontent.com/2073526/159273517-ea9244b1-e438-48a2-84e5-5b96d8415067.png)

lucasmrod commented:
Added the following change:

![Screen Shot 2022-03-21 at 11 24 07](https://user-images.githubusercontent.com/2073526/159318162-787a51a4-c46a-4d96-a2c1-3d020cda3e27.png)

It takes ~1m for Fleet to register the device auth token (depending on `distributed_interval` configuration).
So I added this "Initializing..." state so that the user won't get 500s when visiting the URL right after install.

lucasmrod commented:
Orbit won't auto-update users on macOS to the Fleet Desktop application.
Instead, users that want to try out Fleet Desktop will have to re-generate a package with `fleetctl [...] --fleet-desktop [...]` and then re-install.

@noahtalerman @zwass

zwass commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -460,6 +565,42 @@ func loadOrGenerateToken(rootDir string) (string, error) {
 	}
 }
 
+func killProcessByName(name string) error {
+	foundProcess, err := getProcessByName(name)
+	if err != nil {
+		return fmt.Errorf("get process: %w", err)
+	}
+	if err := foundProcess.Kill(); err != nil {
+		return fmt.Errorf("kill process %d: %w", foundProcess.Pid, err)
+	}
+	return nil
+}
+
+var errProcessNotFound = errors.New("process not found")
+
+func getProcessByName(name string) (*gopsutil_process.Process, error) {
```
I wonder if we'll want to refactor these into some other package for general use at a later time. Not necessary now IMO.

zwass reviewed (approved):
Let's do it!

noahtalerman commented:
> I added this "Initializing..." state so that the user won't get 500s when visiting the URL right after install

@lucasmrod great catch. Your "Initializing..." copy is great for now. 

Now that we're aware that this "loading state" is necessary, the product team can help design these states when wireframing future iterations.

zwass commented:
@noahtalerman I tried out the "template icon" and I think it looks great!

PR: https://github.com/fleetdm/fleet/pull/4784
